### PR TITLE
[PfemFluid] heat source due to friction

### PR DESCRIPTION
**Pull request by AFranci**
This PR introduces the heat source due to friction in the thermal coupled solver. This heat source is considered when using the new pfem_fluid_thermal_solver.

### CODE REVIEW

File: applications/PfemFluidDynamicsApplication/custom_elements/two_step_updated_lagrangian_V_P_implicit_element.cpp [hunk 1]
```diff
@@ -223,19 +223,16 @@ namespace Kratos
...
-                                                                                    std::vector<double> &rOutput,
-                                                                                    const ProcessInfo &rCurrentProcessInfo)
+                                                                                     std::vector<double> &rOutput,
+                                                                                     const ProcessInfo &rCurrentProcessInfo)
   {
-    if (rVariable == EQ_STRAIN_RATE)
-    {
-      rOutput[0] = this->GetValue(EQ_STRAIN_RATE);
-    }
+    rOutput[0] = this->GetValue(rVariable);
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_elements/two_step_updated_lagrangian_V_P_implicit_element.cpp]: In order this to be **safe** you have to be sure to have `SetValue` its value before... If you are not sure you should use `this->HasValue()`

File: applications/PfemFluidDynamicsApplication/custom_elements/two_step_updated_lagrangian_V_P_implicit_fluid_DEM_coupling_element.cpp [hunk 2]
```diff
@@ -44,6 +44,8 @@ namespace Kratos
   template <unsigned int TDim>
   void TwoStepUpdatedLagrangianVPImplicitFluidDEMcouplingElement<TDim>::InitializeSolutionStep(const ProcessInfo &rCurrentProcessInfo)
   {
+    KRATOS_TRY;
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_elements/two_step_updated_lagrangian_V_P_implicit_fluid_DEM_coupling_element.cpp]: If it's empty you can remove the method?

File: applications/PfemFluidDynamicsApplication/custom_elements/two_step_updated_lagrangian_V_P_implicit_fluid_DEM_coupling_element.cpp [hunk 3]
```diff
@@ -355,7 +357,7 @@ namespace Kratos
                                                                                            const double BoundRHSCoeffDev)
   {
     GeometryType &rGeom = this->GetGeometry();
-    //const SizeType NumNodes = rGeom.PointsNumber();
+    // const SizeType NumNodes = rGeom.PointsNumber();
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_elements/two_step_updated_lagrangian_V_P_implicit_fluid_DEM_coupling_element.cpp]: remove code?

File: applications/PfemFluidDynamicsApplication/custom_elements/two_step_updated_lagrangian_V_P_implicit_fluid_FIC_element.cpp [hunk 4]
```diff
@@ -55,6 +55,8 @@ namespace Kratos
   template <unsigned int TDim>
   void TwoStepUpdatedLagrangianVPImplicitFluidFicElement<TDim>::InitializeSolutionStep(const ProcessInfo &rCurrentProcessInfo)
   {
+    KRATOS_TRY;
+    KRATOS_CATCH("");
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_elements/two_step_updated_lagrangian_V_P_implicit_fluid_FIC_element.cpp]: remove method if empty?

File: applications/PfemFluidDynamicsApplication/custom_elements/two_step_updated_lagrangian_V_P_implicit_fluid_FIC_element.cpp [hunk 5]
```diff
@@ -142,9 +144,9 @@ namespace Kratos
     const auto &r_geometry = this->GetGeometry();
     const SizeType dimension = r_geometry.WorkingSpaceDimension();
 
-    //WARNING THIS MUST BE REMOVED ASAP
+    // WARNING THIS MUST BE REMOVED ASAP
     const_cast<TwoStepUpdatedLagrangianVPImplicitFluidFicElement<TDim> *>(this)->mpConstitutiveLaw = const_cast<TwoStepUpdatedLagrangianVPImplicitFluidFicElement<TDim> *>(this)->GetProperties().GetValue(CONSTITUTIVE_LAW);
-    //mpConstitutiveLaw = this->GetProperties().GetValue(CONSTITUTIVE_LAW);
+    // mpConstitutiveLaw = this->GetProperties().GetValue(CONSTITUTIVE_LAW);
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_elements/two_step_updated_lagrangian_V_P_implicit_fluid_FIC_element.cpp:135]: remove unused code

File: applications/PfemFluidDynamicsApplication/custom_elements/two_step_updated_lagrangian_V_P_implicit_fluid_FIC_element.cpp [hunk 6]
```diff
@@ -930,7 +932,7 @@ namespace Kratos
             laplacianRHSi += StabLaplacianWeight * rDN_DX(i, d) * OldPressureGradient[d];
           }
           rRightHandSideVector[i] += -laplacianRHSi;
-          //NewRhsLaplacian[i] += -laplacianRHSi;
+          // NewRhsLaplacian[i] += -laplacianRHSi;
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_elements/two_step_updated_lagrangian_V_P_implicit_fluid_FIC_element.cpp:921]: unused code

File: applications/PfemFluidDynamicsApplication/custom_elements/two_step_updated_lagrangian_V_P_implicit_fluid_PSPG_element.cpp [hunk 7]
```diff
@@ -55,6 +55,8 @@ namespace Kratos
   template <unsigned int TDim>
   void TwoStepUpdatedLagrangianVPImplicitFluidPspgElement<TDim>::InitializeSolutionStep(const ProcessInfo &rCurrentProcessInfo)
   {
+    KRATOS_TRY;
+    KRATOS_CATCH("");
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_elements/two_step_updated_lagrangian_V_P_implicit_fluid_PSPG_element.cpp]: idem

File: applications/PfemFluidDynamicsApplication/custom_elements/two_step_updated_lagrangian_V_P_implicit_fluid_PSPG_element.cpp [hunk 8]
```diff
@@ -353,7 +355,7 @@ namespace Kratos
     this->CalcMeanVelocityNorm(MeanVelocity, 0);
 
     // // Tau Fic
-    //Tau = (ElemSize * ElemSize * DeltaTime) / (Density * MeanVelocity * DeltaTime * ElemSize + Density * ElemSize * ElemSize + 8.0 * Viscosity * DeltaTime);
+    // Tau = (ElemSize * ElemSize * DeltaTime) / (Density * MeanVelocity * DeltaTime * ElemSize + Density * ElemSize * ElemSize + 8.0 * Viscosity * DeltaTime);
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_elements/two_step_updated_lagrangian_V_P_implicit_fluid_PSPG_element.cpp:344]: unused

File: applications/PfemFluidDynamicsApplication/custom_elements/two_step_updated_lagrangian_V_P_implicit_fluid_PSPG_element.cpp [hunk 9]
```diff
@@ -471,8 +473,8 @@ namespace Kratos
       // double acc_Y = (this->GetGeometry()[j].FastGetSolutionStepValue(VELOCITY_Y,0) - this->GetGeometry()[j].FastGetSolutionStepValue(VELOCITY_Y,1)) / TimeStep;
 
       RHSi += rDN_DX(i, 0) * rN[j] * this->GetGeometry()[j].FastGetSolutionStepValue(ACCELERATION_X, 0) + rDN_DX(i, 1) * rN[j] * this->GetGeometry()[j].FastGetSolutionStepValue(ACCELERATION_Y, 0);
-      //RHSi += Tau * Density * (rDN_DX(i, 0) + rDN_DX(i, 1)) * rN[j] * (acc_X + acc_Y) ;
-      // RHSi +=  rN[i] *(rDN_DX(j, 0)*this->GetGeometry()[j].FastGetSolutionStepValue(VELOCITY_X,0) + rDN_DX(j, 1)*this->GetGeometry()[j].FastGetSolutionStepValue(VELOCITY_Y,0));
+      // RHSi += Tau * Density * (rDN_DX(i, 0) + rDN_DX(i, 1)) * rN[j] * (acc_X + acc_Y) ;
+      //  RHSi +=  rN[i] *(rDN_DX(j, 0)*this->GetGeometry()[j].FastGetSolutionStepValue(VELOCITY_X,0) + rDN_DX(j, 1)*this->GetGeometry()[j].FastGetSolutionStepValue(VELOCITY_Y,0));
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_elements/two_step_updated_lagrangian_V_P_implicit_fluid_PSPG_element.cpp:463]: unused

File: applications/PfemFluidDynamicsApplication/custom_elements/two_step_updated_lagrangian_V_P_implicit_fluid_element.cpp [hunk 10]
```diff
@@ -578,6 +580,14 @@ namespace Kratos
       }
     }
 
+    const double volumetric_strain = (rElementalVariables.SpatialDefRate[0] + rElementalVariables.SpatialDefRate[1]) * 0.5;
+
+    double mechanicalDissipation = rElementalVariables.UpdatedDeviatoricCauchyStress[0] * (rElementalVariables.SpatialDefRate[0] - volumetric_strain) +
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_elements/two_step_updated_lagrangian_V_P_implicit_fluid_element.cpp:571]: const?

File: applications/PfemFluidDynamicsApplication/custom_elements/two_step_updated_lagrangian_V_P_implicit_fluid_element.cpp [hunk 11]
```diff
@@ -654,6 +664,17 @@ namespace Kratos
       }
     }
 
+    const double volumetric_strain = (rElementalVariables.SpatialDefRate[0] + rElementalVariables.SpatialDefRate[1] + rElementalVariables.SpatialDefRate[2]) / 3.0;
+
+    double mechanicalDissipation = rElementalVariables.UpdatedDeviatoricCauchyStress[0] * (rElementalVariables.SpatialDefRate[0] - volumetric_strain) +
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_elements/two_step_updated_lagrangian_V_P_implicit_fluid_element.cpp:655]: const

File: applications/PfemFluidDynamicsApplication/custom_elements/two_step_updated_lagrangian_V_P_implicit_solid_element.cpp [hunk 12]
```diff
@@ -388,6 +388,14 @@ namespace Kratos
     this->mMaterialDeviatoricCoefficient = DeviatoricCoeff;
     this->mMaterialVolumetricCoefficient = VolumetricCoeff;
     this->mMaterialDensity = Density;
+
+    const double volumetric_strain = (rElementalVariables.SpatialDefRate[0] + rElementalVariables.SpatialDefRate[1]) * 0.5;
+
+    double mechanicalDissipation = rElementalVariables.UpdatedDeviatoricCauchyStress[0] * (rElementalVariables.SpatialDefRate[0] - volumetric_strain) +
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_elements/two_step_updated_lagrangian_V_P_implicit_solid_element.cpp:389]: const // besides I see some code repetition between elements. Maybe you can create a PFEMFluidUtilities to include all this shared methods (see ConstitutiveLawsUtilities)

File: applications/PfemFluidDynamicsApplication/custom_elements/two_step_updated_lagrangian_V_P_implicit_solid_element.cpp [hunk 13]
```diff
@@ -456,6 +464,17 @@ namespace Kratos
     this->mMaterialDeviatoricCoefficient = DeviatoricCoeff;
     this->mMaterialVolumetricCoefficient = VolumetricCoeff;
     this->mMaterialDensity = Density;
+
+    const double volumetric_strain = (rElementalVariables.SpatialDefRate[0] + rElementalVariables.SpatialDefRate[1] + rElementalVariables.SpatialDefRate[2]) / 3.0;
+
+    double mechanicalDissipation = rElementalVariables.UpdatedDeviatoricCauchyStress[0] * (rElementalVariables.SpatialDefRate[0] - volumetric_strain) +
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_elements/two_step_updated_lagrangian_V_P_implicit_solid_element.cpp:465]: again

File: applications/PfemFluidDynamicsApplication/custom_elements/updated_lagrangian_V_implicit_solid_element.cpp [hunk 14]
```diff
@@ -121,6 +121,14 @@ void UpdatedLagrangianVImplicitSolidElement<2>::CalcElasticPlasticCauchySplitted
 
     this->mUpdatedTotalCauchyStress[g] = rElementalVariables.UpdatedTotalCauchyStress;
     this->mUpdatedDeviatoricCauchyStress[g] = rElementalVariables.UpdatedDeviatoricCauchyStress;
+
+    const double volumetric_strain = (rElementalVariables.SpatialDefRate[0] + rElementalVariables.SpatialDefRate[1]) * 0.5;
+
+    double mechanicalDissipation = rElementalVariables.UpdatedDeviatoricCauchyStress[0] * (rElementalVariables.SpatialDefRate[0] - volumetric_strain) +
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_elements/updated_lagrangian_V_implicit_solid_element.cpp:127]: aand again

File: applications/PfemFluidDynamicsApplication/custom_elements/updated_lagrangian_V_implicit_solid_element.cpp [hunk 15]
```diff
@@ -189,6 +197,17 @@ void UpdatedLagrangianVImplicitSolidElement<3>::CalcElasticPlasticCauchySplitted
 
     this->mUpdatedTotalCauchyStress[g] = rElementalVariables.UpdatedTotalCauchyStress;
     this->mUpdatedDeviatoricCauchyStress[g] = rElementalVariables.UpdatedDeviatoricCauchyStress;
+
+    const double volumetric_strain = (rElementalVariables.SpatialDefRate[0] + rElementalVariables.SpatialDefRate[1] + rElementalVariables.SpatialDefRate[2]) / 3.0;
+
+    double mechanicalDissipation = rElementalVariables.UpdatedDeviatoricCauchyStress[0] * (rElementalVariables.SpatialDefRate[0] - volumetric_strain) +
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_elements/updated_lagrangian_V_implicit_solid_element.cpp:203]: here too

File: applications/PfemFluidDynamicsApplication/custom_strategies/strategies/two_step_v_p_strategy.h [hunk 16]
```diff
@@ -269,8 +271,8 @@ namespace Kratos
         if (it == maxNonLinearIterations - 1 || ((continuityConverged && momentumConverged) && it > 2))
         {
 
-          //double tensilStressSign = 1.0;
-          // ComputeErrorL2Norm(tensilStressSign);
+          // double tensilStressSign = 1.0;
+          //  ComputeErrorL2Norm(tensilStressSign);
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_strategies/strategies/two_step_v_p_strategy.h:275]: unused

**Review by AlejandroCornejo — CHANGES_REQUESTED** [review]

File: applications/PfemFluidDynamicsApplication/custom_strategies/strategies/two_step_v_p_strategy.h (hunk 16 above)
Comment (AFranci) [reply applications/PfemFluidDynamicsApplication/custom_strategies/strategies/two_step_v_p_strategy.h:275]: thanks as usual for the comments @AlejandroCornejo 
I have made the changes requested. I'll make the utility for constitutive law in a separate PR, if it's ok for you.

File: applications/PfemFluidDynamicsApplication/custom_elements/two_step_updated_lagrangian_V_P_implicit_fluid_DEM_coupling_element.cpp [hunk 17]
```diff
@@ -34,27 +34,6 @@ namespace Kratos
     return Element::Pointer(new TwoStepUpdatedLagrangianVPImplicitFluidDEMcouplingElement(NewElement));
   }
 
-  template <unsigned int TDim>
-  void TwoStepUpdatedLagrangianVPImplicitFluidDEMcouplingElement<TDim>::Initialize(const ProcessInfo &rCurrentProcessInfo)
-  {
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_elements/two_step_updated_lagrangian_V_P_implicit_fluid_DEM_coupling_element.cpp:39]: nicee

File: applications/PfemFluidDynamicsApplication/custom_elements/two_step_updated_lagrangian_V_P_implicit_fluid_FIC_element.cpp [hunk 18]
```diff
@@ -52,20 +52,6 @@ namespace Kratos
...
 
-  template <unsigned int TDim>
-  void TwoStepUpdatedLagrangianVPImplicitFluidFicElement<TDim>::InitializeSolutionStep(const ProcessInfo &rCurrentProcessInfo)
-  {
-    KRATOS_TRY;
-    KRATOS_CATCH("");
-  }
-
-  template <unsigned int TDim>
-  void TwoStepUpdatedLagrangianVPImplicitFluidFicElement<TDim>::InitializeNonLinearIteration(const ProcessInfo &rCurrentProcessInfo)
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_elements/two_step_updated_lagrangian_V_P_implicit_fluid_FIC_element.cpp:61]: cleaaninggg

File: applications/PfemFluidDynamicsApplication/custom_elements/two_step_updated_lagrangian_V_P_implicit_solid_element.cpp [hunk 19]
```diff
@@ -211,7 +206,7 @@ namespace Kratos
     const auto &r_geometry = this->GetGeometry();
     const SizeType dimension = r_geometry.WorkingSpaceDimension();
 
-    //WARNING THIS MUST BE REMOVED ASAP
+    // WARNING THIS MUST BE REMOVED ASAP
     const_cast<TwoStepUpdatedLagrangianVPImplicitSolidElement<TDim> *>(this)->mpConstitutiveLaw = const_cast<TwoStepUpdatedLagrangianVPImplicitSolidElement<TDim> *>(this)->GetProperties().GetValue(CONSTITUTIVE_LAW);
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_elements/two_step_updated_lagrangian_V_P_implicit_solid_element.cpp:210]: c++ nightmare

**Review by AlejandroCornejo — APPROVED** [review]

**Pull request merged**